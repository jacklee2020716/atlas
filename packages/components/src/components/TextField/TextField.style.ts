import { css } from "@emotion/core"
import { typography, colors, spacing } from "./../../theme"
import { icon } from "@fortawesome/fontawesome-svg-core"

export type TextFieldStyleProps = {
  disabled?: boolean
  focus?: boolean
  error?: boolean
  isActive?: boolean
  iconPosition?: "right" | "left"
}

export let makeStyles = ({
  disabled = false,
  focus = false,
  error = false,
  isActive = false,
  iconPosition = "right"
}: TextFieldStyleProps) => {

  const borderColor = disabled ? colors.gray[200] :
    error ? colors.error :
    focus ? colors.blue[500] :
    isActive ? colors.gray[200] : colors.gray[400]

  return {
    wrapper: css`
      display: block;
      font-family: ${typography.fonts.base};
    `,
    container: css`
      position: relative;
      min-width: 250px;
      height: 50px;
      display: inline-flex;
      cursor: ${disabled ? "not-allowed" : "default"};
    `,
    border: css`
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      border: 1px solid ${borderColor};
      display: flex;
      align-items: center;
      justify-content: left;
    `,
    label: css`
      color: ${error ? colors.error : colors.gray[400]};
      padding: 0 ${spacing.s};
      ${icon ? `padding-${iconPosition}: ${spacing.xxxxl};` : ""}
      background-color: ${colors.black};
      transition: all 0.1s linear;
    `,
    input: css`
      display: none;
      width: 100%;
      margin: 0 ${spacing.s};
      ${icon ? `margin-${iconPosition}: ${spacing.xxxxl};` : ""}
      background: none;
      border: none;
      color: ${colors.white};
      outline: none;
    `,
    icon: css`
      color: ${colors.gray[300]};
      font-size: ${typography.sizes.icon.xlarge};
      position: absolute;
      top: ${spacing.s};
      ${iconPosition}: ${spacing.s};
    `,
    helper: css`
      color: ${error ? colors.error : colors.gray[400]};
      font-size: ${typography.sizes.caption};
      margin: ${spacing.xxs} ${spacing.xs};
    `
  }
}
