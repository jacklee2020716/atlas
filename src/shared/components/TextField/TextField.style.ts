import { makeStyles, StyleFn } from '../../utils'
import { colors, spacing, typography } from './../../theme'

export type TextFieldStyleProps = {
  disabled?: boolean
  focus?: boolean
  error?: boolean
  isActive?: boolean
}

const FIELD_WIDTH = '250px'
const wrapper: StyleFn = () => ({
  display: 'block',
  maxWidth: FIELD_WIDTH,
  fontFamily: typography.fonts.base,
})

const container: StyleFn = (_, { disabled }) => ({
  position: 'relative',
  width: '100%',
  height: '48px',
  display: 'inline-flex',
  cursor: disabled ? 'not-allowed' : 'default',
})

const border: StyleFn = () => ({
  position: 'absolute',
  top: 0,
  left: 0,
  right: 0,
  bottom: 0,
  borderWidth: '1px',
  borderStyle: 'solid',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'left',
})
const borderColor: StyleFn = (styles, { disabled, error, isActive, focus }) => {
  const borderColor = disabled
    ? colors.gray[200]
    : error
    ? colors.error
    : focus
    ? colors.blue[500]
    : isActive
    ? colors.gray[200]
    : colors.gray[400]

  return {
    ...styles,
    borderColor,
  }
}
const label: StyleFn = (_, { error }) => ({
  color: error ? colors.error : colors.gray[400],
  padding: `0 ${spacing.s}`,
  backgroundColor: colors.black,
  fontSize: typography.sizes.body2,
  transition: `all 0.1s linear`,
})
const input: StyleFn = () => ({
  display: 'none',
  width: '100%',
  margin: `0 ${spacing.s}`,
  background: 'none',
  border: 'none',
  outline: 'none',
  color: colors.white,
  fontSize: typography.sizes.body2,
  padding: `5px 0`,
})
const helper: StyleFn = (_, { error }) => ({
  color: error ? colors.error : colors.gray[400],
})

export const useCSS = ({ disabled = false, focus = false, error = false, isActive = false }: TextFieldStyleProps) => {
  const props = { disabled, focus, error, isActive }
  return {
    wrapper: makeStyles([wrapper])(props),
    container: makeStyles([container])(props),
    border: makeStyles([border, borderColor])(props),
    label: makeStyles([label])(props),
    input: makeStyles([input])(props),
    helper: makeStyles([helper])(props),
  }
}
