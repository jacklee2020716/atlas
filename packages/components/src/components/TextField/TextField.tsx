import React, { useState, useRef, useEffect } from "react";
import { useCSS, TextFieldStyleProps } from "./TextField.style";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { spacing } from "./../../theme";

type TextFieldProps = {
	label: string;
	helper?: string;
	value?: string;
	icon?: IconProp;
	onChange?: (e: React.ChangeEvent) => void;
} & TextFieldStyleProps;

export default function TextField({
	label,
	helper = "",
	value = "",
	icon = null,
	disabled = false,
	onChange,
	...styleProps
}: TextFieldProps) {
	const inputRef = useRef(null);
	const [isActive, setIsActive] = useState(!!value);
	const [inputTextValue, setInputTextValue] = useState(value);
	const styles = useCSS({ isActive, disabled, ...styleProps });

	useEffect(() => {
		if (isActive) {
			inputRef.current.focus();
		} else {
			inputRef.current.blur();
		}
	}, [isActive, inputRef]);

	function onTextFieldClick() {
		if (!disabled) setIsActive(true);
	}

	function onInputTextBlur() {
		setIsActive(false);
	}

	// FIXME: add correct typing to event, see here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12239
	function onInputTextChange(event: any): void {
		if (!disabled) setInputTextValue(event.currentTarget.value);
	}

	return (
		<div css={styles.wrapper}>
			<div css={styles.container} onClick={onTextFieldClick}>
				<div css={styles.border}>
					<div
						css={styles.label}
						style={
							!inputTextValue && !isActive
								? {}
								: {
										position: "absolute",
										top: "-8px",
										left: "5px",
										fontSize: "0.7rem",
										padding: `0 ${spacing.xs}`,
								  }
						}
					>
						{label}
					</div>
					<input
						css={styles.input}
						style={{ display: !!inputTextValue || isActive ? "block" : "none" }}
						ref={inputRef}
						type="text"
						value={inputTextValue}
						onChange={disabled ? null : onInputTextChange}
						onBlur={onInputTextBlur}
					/>
					{!!icon && <FontAwesomeIcon icon={icon} css={styles.icon} />}
				</div>
			</div>
			{!!helper && <p css={styles.helper}>{helper}</p>}
		</div>
	);
}
