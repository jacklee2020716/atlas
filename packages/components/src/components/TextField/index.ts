import { memo } from "react"
import TextField from "./TextField"
export default memo(TextField)
