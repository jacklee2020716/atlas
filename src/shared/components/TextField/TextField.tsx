import React, { useEffect, useRef, useState } from 'react'
import { TextFieldStyleProps, useCSS } from './TextField.style'
import { spacing } from './../../theme'

type TextFieldProps = {
  label: string
  helper?: string
  value?: string
  onChange?: (e: React.ChangeEvent) => void
} & TextFieldStyleProps

export default function TextField({ label, helper = '', value = '', disabled = false, ...styleProps }: TextFieldProps) {
  const inputRef = useRef<HTMLInputElement>(null)
  const [isActive, setIsActive] = useState(!!value)
  const [inputTextValue, setInputTextValue] = useState(value)
  const styles = useCSS({ isActive, disabled, ...styleProps })

  useEffect(() => {
    if (inputRef.current != null) {
      if (isActive) {
        inputRef.current.focus()
      } else {
        inputRef.current.blur()
      }
    }
  }, [isActive, inputRef])

  function onTextFieldClick() {
    if (!disabled) setIsActive(true)
  }

  function onInputTextBlur() {
    setIsActive(false)
  }

  // FIXME: add correct typing to event, see here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/12239
  function onInputTextChange(event: any): void {
    if (!disabled) setInputTextValue(event.currentTarget.value)
  }

  return (
    <div css={styles.wrapper}>
      <div css={styles.container} onClick={onTextFieldClick}>
        <div css={styles.border}>
          <div
            css={styles.label}
            style={
              !inputTextValue && !isActive
                ? {}
                : {
                    position: 'absolute',
                    top: '-8px',
                    left: '5px',
                    fontSize: '0.7rem',
                    padding: `0 ${spacing.xs}`,
                  }
            }
          >
            {label}
          </div>
          <input
            css={styles.input}
            style={{ display: !!inputTextValue || isActive ? 'block' : 'none' }}
            ref={inputRef}
            type="text"
            value={inputTextValue}
            onChange={onInputTextChange}
            onBlur={onInputTextBlur}
            disabled={disabled}
          />
        </div>
      </div>
      {!!helper && <p css={styles.helper}>{helper}</p>}
    </div>
  )
}
