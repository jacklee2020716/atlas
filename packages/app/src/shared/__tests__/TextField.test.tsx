import React from 'react'
import { mount } from 'enzyme'
import TextField from '@/shared/components/TextField'
describe('TextField component', () => {
  it('Should render Text Field correctly', () => {
    expect(mount(<TextField label="Test TextField" />)).toMatchSnapshot()
  })
})
