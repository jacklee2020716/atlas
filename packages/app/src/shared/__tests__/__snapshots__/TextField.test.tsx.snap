// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component Should render Text Field correctly 1`] = `ReactWrapper {}`;
